### DISCUSSION TITLE
About Simulation Time

### DISCUSSION THREAD

**Discussion by iftahnaf** [citros-garden/.github#24] [Ideas]
Hi all,

I'm starting a discussion about timing and fixed / variant time steps in ROS-based simulators.

# General Information

An autonomous robotics simulator has 4 components:

1. **Physics model** which includes the robot configuration and dynamics equations and the world model - winds, gravity, sun, etc.
2. **Sensors models** that simulate the sensors' reading with correlation to their specifications.
3. **Controller** low-level control which commands the robot actuators.
4. **Guidance / Path Planner**  responsible for the trajectory of the robot, sends setpoints to the controller.

The data flow can be described as follows:
```mermaid
graph TB
    subgraph Simulator
        A[Physics Model] --> B[Sensors Models]
        B --> C[Guidance / Path Planner]
        C --> D[Controller]
    end
    D --> A
```

*In some configurations, we can use an `ideal` model, for example, no sensors models, the actual state is being sent to the guidance without noises.*

# Fixed-Time Step
## Physical Model Time
In some available robotics simulators (Simulink, Gazebo) in most cases the time step for solving the differential equations(for state calculation) is fixed, meaning each step of the physical model the time is known and deterministic (almost?).

## Sensors / Controller / Guidance
As a consequence of the deterministic fixed physics time step, the other simulator elements can assume that the physics advancing in fixed time steps, therefore they can also use a fixed time step at their ODE solution. The small variance in the time steps is compensated by the closed-loop controller, which closes the errors generated by this.

# Current CITROS-Garden Demos Configuration
In our Gazebo-based physics simulators (drone), we based our time sync on the [Gazebo - PX4](https://docs.px4.io/main/en/simulation/) lockstep sync - which is very deterministic sync between the controller, sensors and the physical simulator. When we run the `offboard` node, we just assume that it runs close enough to the PX4 - Gazebo time (which is supposed to be close to "real-time") and it works.

In our Python-based physics simulators, we have the following configuration:

```mermaid

graph TB
    subgraph Simulator
        A[Scheduler] --> B[Physics ]
        B --> C[Controller]
    end
    C --> B

```

Where the scheduler triggers the physics, which triggers the controller (callback from the physics topic).
In terms of timing, for each physics step, there is a controller step, and both ODEs are solved by a fixed time step.

# Open Discussion

* In some cases (XPlane🤮) the physics time step is based on the graphical step (image rendering) which is very not deterministic, therefore if the closed-loop controller is not robust enough (it's not), the simulation will not work. Do we want to address such issues in the Garden project? 
* Is it possible to do a variant-time step simulation, meaning get the precise physics time step and use it in the other elements ' ODEs (variant time-step), and do we even want to do it?

**Comment by iftahnaf** [comment]
I opened this discussion after talking with @vovacooper, @noamoon @lulaviv @BoazonG @orrlulavspace @gtep96 @shalevm  would love to hear your opinion about this.

**Reply by vovacooper** [reply]
@freddy33

**Comment by iftahnaf** [comment]
@vovacooper :
* i would expect to see another garden example without the scheduler
```mermaid

graph TB
    subgraph Simulator        
        B --> C[Controller]
    end
    C --> B

```
* The incentive to connect the **physical world** time to the **game time (simulation)** is the visual part that has to look realistic for the human eye.

**Comment by iftahnaf** [comment]
@vovacooper :


# Dynamic-Time Step - Game engine deriven simulation (xplain/Unity/Unreal/ etc...)
most of the game engines work like this:
<img width="703" alt="image" src="https://github.com/citros-garden/.github/assets/2060520/c6fc1645-3ebe-478d-b801-a0428d1a5d40">
here you can see on update function is being called not in exact dt each frame, but it depend on the time it took to render, which can very very much from frame to frame.
